examples/list.bend

```diff
@@ -0,0 +1,169 @@
+##############################
+# Author: Ematth, 2024
+##############################
+### Singly-Linked List Type Definition: ###
+
+# type List:
+#   Cons { head, ~tail }
+#   Nil
+
+###########################################
+
+# List clear:
+# List l -> list l
+# clears all elements from list l.
+# (this might be completely useless...)
+List/clear = @l
+  []
+
+# List concat:
+# List l -> List l
+# combines two lists (l1, l2) from left to right.
+List/concat = @l1 @l2
+  match l1 {
+    List/Cons: (List/Cons l1.head (List/concat l1.tail l2))
+    List/Nil: l2
+  }
+
+# List add_front:
+# List l -> List l
+# adds a non-List element e to the front of list l.
+List/add_front = @l @e
+  match l {
+    List/Cons: (List/Cons e l)
+    List/Nil: (List/Cons e List/Nil)
+  }
+
+# List append (add_back):
+# List l -> List l
+# adds a non-list element e to the back of list l.
+List/append = @l @e
+  (List/concat l (List/Cons e List/Nil))
+
+# List reverse:
+# List l -> List l
+# reverses the order of elements in list l.
+List/reverse/aux = @acc @l
+  match l {
+    List/Nil: acc
+    List/Cons: (List/reverse/aux (List/Cons l.head acc) l.tail)
+  }
+List/reverse = @l
+  (List/reverse/aux [] l)
+
+# List length:
+# List l -> uint
+# returns the number of elements in list l.
+List/len/aux = @acc @l
+  match l {
+    List/Nil: acc
+    List/Cons: (List/len/aux (+ acc 1) l.tail)
+  }
+List/len = @l
+  (List/len/aux 0 l)
+
+# List count:
+# List l -> Some s -> uint
```

Comment:
Here `s` is necessarily a number.
To make it generic you could pass a comparison function

Reply:
This applies to all the functions where == or another numeric operation with the element is done

Reply:
That would be me assuming == working as a general comparison. I'll reserve it for uint for now and generalize comparisons later.